oura-library/src/main/kotlin/org/radarbase/oura/request/OuraRequestGenerator.kt

```diff
@@ -0,0 +1,193 @@
+                        }
+
+                        val endDate = user.endDate
+                        if (endDate <= startOffset) return@flatMap emptySequence()
+                        val endTime = (startOffset + defaultQueryRange).coerceAtMost(endDate)
+                        route.generateRequests(user, startOffset, endTime, max / routes.size)
+                    }
+                    .takeWhile { !shouldBackoff }
+    }
+
+    override fun requests(route: Route, user: User, max: Int): Sequence<RestRequest> {
+        return if (user.ready()) {
+//                        val offsets: Offsets? = offsetPersistenceFactory.read(user.versionedId)
+                val offsets = null
+                val backfillLimit = Instant.now().minus(route.maxBackfillPeriod())
+//                        val startDate = userRepository.getBackfillStartDate(user)
+                val startDate = user.startDate
+                var startOffset: Instant = if (offsets == null) {
+                    logger.debug("No offsets found for $user, using the start date.")
+                    startDate
+                } else {
+                    logger.debug("Offsets found in persistence.")
+                    startDate
+//                            offsets.offsetsMap.getOrDefault(
+//                                    UserRoute(user.versionedId, route.toString()), startDate
+//                            ).coerceAtLeast(startDate)
+                }
+
+                if (startOffset <= backfillLimit) {
+                    // the start date is before the backfill limits
+                    logger.warn(
+                            "Backfill limit exceeded for $user and $route. " +
+                                    "Resetting to earliest allowed start offset."
+                    )
+                    startOffset = backfillLimit.plus(Duration.ofDays(2))
+                }
+
+                val endDate = user.endDate
+                if (endDate <= startOffset) return@flatMap emptySequence()
+                val endTime = (startOffset + defaultQueryRange).coerceAtMost(endDate)
+                return route.generateRequests(user, startOffset, endTime, max / routes.size)
+        } else emptySequence()
+    }
+    
+
+    override fun requestSuccessful(request: RestRequest, response: Response) {
+        logger.debug("Request successful: {}. Writing to offsets...", request.request)
+//        offsetPersistenceFactory.add(
+//                Path.of(request.user.versionedId), UserRouteOffset(
+//                request.user.versionedId,
+//                request.route.toString(),
+//                request.endDate
+//        )
+//        )
+    }
+
+    override fun requestFailed(request: RestRequest, response: Response) {
+        when (response.code) {
+            429 -> {
+                logger.info("Too many requests, rate limit reached. Backing off...")
+                nextRequestTime = Instant.now() + BACK_OFF_TIME
```

Comment:
Check nextRequestTime is before now.

Reply:
Checked here: https://github.com/RADAR-base/RADAR-REST-Connector/pull/102/files#diff-3c67160eed894e7b317044e0b11882d5012d73ea78f29cca0c8baa526e033594R40